File: mcp-server/src/tools/sketch-tools.ts
import { z } from 'zod';
import type { NamedPipeClient } from '../transport/named-pipe-client.js';

// ── Response shapes from C# bridge ─────────────────────────────

export interface SwSketchEntityInfo {
  type: string;   // "Line" | "Circle" | "Rectangle" | "Arc"
  id: string;     // segment identifier
}

// ── Zod input schemas ───────────────────────────────────────────

export const InsertSketchSchema = z
  .object({})
  .describe('Open a new sketch on the currently selected plane or face');

export const FinishSketchSchema = z
  .object({})
  .describe('Close (finish) the currently open sketch');

export const AddPointSchema = z.object({
  x: z.number().describe('Point X coordinate (meters)'),
  y: z.number().describe('Point Y coordinate (meters)'),
});

export const AddEllipseSchema = z.object({
  cx: z.number().describe('Ellipse center X coordinate (meters)'),
  cy: z.number().describe('Ellipse center Y coordinate (meters)'),
  majorX: z.number().describe('Major-axis point X coordinate (meters)'),
  majorY: z.number().describe('Major-axis point Y coordinate (meters)'),
  minorX: z.number().describe('Minor-axis point X coordinate (meters)'),
  minorY: z.number().describe('Minor-axis point Y coordinate (meters)'),
});

export const AddPolygonSchema = z.object({
  cx: z.number().describe('Polygon center X coordinate (meters)'),
  cy: z.number().describe('Polygon center Y coordinate (meters)'),
  x: z.number().describe('A point on the polygon perimeter X coordinate (meters)'),
  y: z.number().describe('A point on the polygon perimeter Y coordinate (meters)'),
  sides: z.number().int().min(3).describe('Number of polygon sides'),
  inscribed: z.boolean().default(true).describe('True for inscribed polygon, false for circumscribed polygon'),
});

export const AddTextSchema = z.object({
  x: z.number().describe('Text anchor X coordinate (meters)'),
  y: z.number().describe('Text anchor Y coordinate (meters)'),
  text: z.string().min(1).describe('Sketch text content'),
});

export const AddLineSchema = z.object({
  x1: z.number().describe('Start X coordinate (meters)'),
  y1: z.number().describe('Start Y coordinate (meters)'),
  x2: z.number().describe('End X coordinate (meters)'),
  y2: z.number().describe('End Y coordinate (meters)'),
});

export const AddCircleSchema = z.object({
  cx: z.number().describe('Center X coordinate (meters)'),
  cy: z.number().describe('Center Y coordinate (meters)'),
  radius: z.number().positive().describe('Circle radius (meters)'),
});

export const AddRectangleSchema = z.object({
  x1: z.number().describe('First corner X coordinate (meters)'),
  y1: z.number().describe('First corner Y coordinate (meters)'),
  x2: z.number().describe('Opposite corner X coordinate (meters)'),
  y2: z.number().describe('Opposite corner Y coordinate (meters)'),
});

export const AddArcSchema = z.object({
  cx: z.number().describe('Center X coordinate (meters)'),
  cy: z.number().describe('Center Y coordinate (meters)'),
  x1: z.number().describe('Start point X coordinate (meters)'),
  y1: z.number().describe('Start point Y coordinate (meters)'),
  x2: z.number().describe('End point X coordinate (meters)'),
  y2: z.number().describe('End point Y coordinate (meters)'),
  direction: z
    .number()
    .int()
    .refine((v) => v === 1 || v === -1, { message: 'direction must be 1 (CCW) or -1 (CW)' })
    .describe('Arc direction: 1 = counter-clockwise, -1 = clockwise'),
});

// ── Tool handler functions ──────────────────────────────────────

export async function swInsertSketch(client: NamedPipeClient): Promise<void> {
  const resp = await client.request('sw.sketch.insert');
  if (resp.error) throw new Error(`sw.sketch.insert failed: ${resp.error.message}`);
}

export async function swFinishSketch(client: NamedPipeClient): Promise<void> {
  const resp = await client.request('sw.sketch.finish');
  if (resp.error) throw new Error(`sw.sketch.finish failed: ${resp.error.message}`);
}

export async function swAddPoint(
  client: NamedPipeClient,
  params: z.infer<typeof AddPointSchema>,
): Promise<SwSketchEntityInfo> {
  const resp = await client.request('sw.sketch.add_point', params);
  if (resp.error) throw new Error(`sw.sketch.add_point failed: ${resp.error.message}`);
  return resp.result as SwSketchEntityInfo;
}

export async function swAddEllipse(
  client: NamedPipeClient,
  params: z.infer<typeof AddEllipseSchema>,
): Promise<SwSketchEntityInfo> {
  const resp = await client.request('sw.sketch.add_ellipse', params);
  if (resp.error) throw new Error(`sw.sketch.add_ellipse failed: ${resp.error.message}`);
  return resp.result as SwSketchEntityInfo;
}

export async function swAddPolygon(
  client: NamedPipeClient,
  params: z.infer<typeof AddPolygonSchema>,
): Promise<SwSketchEntityInfo> {
  const resp = await client.request('sw.sketch.add_polygon', params);
  if (resp.error) throw new Error(`sw.sketch.add_polygon failed: ${resp.error.message}`);
  return resp.result as SwSketchEntityInfo;
}

export async function swAddText(
  client: NamedPipeClient,
  params: z.infer<typeof AddTextSchema>,
): Promise<SwSketchEntityInfo> {
  const resp = await client.request('sw.sketch.add_text', params);
  if (resp.error) throw new Error(`sw.sketch.add_text failed: ${resp.error.message}`);
  return resp.result as SwSketchEntityInfo;
}

export async function swAddLine(
  client: NamedPipeClient,
  params: z.infer<typeof AddLineSchema>,
): Promise<SwSketchEntityInfo> {
  const resp = await client.request('sw.sketch.add_line', params);
  if (resp.error) throw new Error(`sw.sketch.add_line failed: ${resp.error.message}`);
  return resp.result as SwSketchEntityInfo;
}

export async function swAddCircle(
  client: NamedPipeClient,
  params: z.infer<typeof AddCircleSchema>,
): Promise<SwSketchEntityInfo> {
  const resp = await client.request('sw.sketch.add_circle', params);
  if (resp.error) throw new Error(`sw.sketch.add_circle failed: ${resp.error.message}`);
  return resp.result as SwSketchEntityInfo;
}

export async function swAddRectangle(
  client: NamedPipeClient,
  params: z.infer<typeof AddRectangleSchema>,
): Promise<SwSketchEntityInfo> {
  const resp = await client.request('sw.sketch.add_rectangle', params);
  if (resp.error) throw new Error(`sw.sketch.add_rectangle failed: ${resp.error.message}`);
  return resp.result as SwSketchEntityInfo;
}

export async function swAddArc(
  client: NamedPipeClient,
  params: z.infer<typeof AddArcSchema>,
): Promise<SwSketchEntityInfo> {
  const resp = await client.request('sw.sketch.add_arc', params);
  if (resp.error) throw new Error(`sw.sketch.add_arc failed: ${resp.error.message}`);
  return resp.result as SwSketchEntityInfo;
}

File: app/SolidWorksMcpApp/Tools/SketchTools.cs
using ModelContextProtocol.Server;
using SolidWorksBridge.SolidWorks;
using System.ComponentModel;
using System.Text.Json;

namespace SolidWorksMcpApp.Tools;

[McpServerToolType]
public class SketchTools(StaDispatcher sta, ISketchService sketch)
{
    [McpServerTool, Description("Open a new sketch on the currently selected datum plane or planar face. This is only valid after exactly one planar face or datum plane has already been selected.")]
    public async Task<string> InsertSketch()
    {
        await sta.InvokeLoggedAsync(nameof(InsertSketch), null, sketch.InsertSketch);
        return "Sketch edit mode started.";
    }

    [McpServerTool, Description("Close (finish) the currently open sketch and return to 3D mode.")]
    public async Task<string> FinishSketch()
    {
        await sta.InvokeLoggedAsync(nameof(FinishSketch), null, sketch.FinishSketch);
        return "Sketch finished.";
    }

    [McpServerTool, Description("Draw a point in the active sketch. Coordinates in meters.")]
    public async Task<string> AddPoint(
        [Description("X coordinate (meters)")] double x,
        [Description("Y coordinate (meters)")] double y)
    {
        var info = await sta.InvokeLoggedAsync(nameof(AddPoint), new { x, y }, () => sketch.AddPoint(x, y));
        return JsonSerializer.Serialize(info);
    }

    [McpServerTool, Description("Draw a line in the active sketch. Coordinates in meters.")]
    public async Task<string> AddLine(
        [Description("Start X (meters)")] double x1,
        [Description("Start Y (meters)")] double y1,
        [Description("End X (meters)")] double x2,
        [Description("End Y (meters)")] double y2)
    {
        var info = await sta.InvokeLoggedAsync(nameof(AddLine), new { x1, y1, x2, y2 }, () => sketch.AddLine(x1, y1, x2, y2));
        return JsonSerializer.Serialize(info);
    }

    [McpServerTool, Description("Draw a circle in the active sketch. Coordinates and radius in meters.")]
    public async Task<string> AddCircle(
        [Description("Center X (meters)")] double cx,
        [Description("Center Y (meters)")] double cy,
        [Description("Radius (meters)")] double radius)
    {
        var info = await sta.InvokeLoggedAsync(nameof(AddCircle), new { cx, cy, radius }, () => sketch.AddCircle(cx, cy, radius));
        return JsonSerializer.Serialize(info);
    }

    [McpServerTool, Description("Draw a corner rectangle in the active sketch. Coordinates in meters.")]
    public async Task<string> AddRectangle(
        [Description("First corner X (meters)")] double x1,
        [Description("First corner Y (meters)")] double y1,
        [Description("Opposite corner X (meters)")] double x2,
        [Description("Opposite corner Y (meters)")] double y2)
    {
        var info = await sta.InvokeLoggedAsync(nameof(AddRectangle), new { x1, y1, x2, y2 }, () => sketch.AddRectangle(x1, y1, x2, y2));
        return JsonSerializer.Serialize(info);
    }

    [McpServerTool, Description("Draw an arc in the active sketch. Coordinates in meters. Direction: 1=CCW, -1=CW.")]
    public async Task<string> AddArc(
        [Description("Center X (meters)")] double cx,
        [Description("Center Y (meters)")] double cy,
        [Description("Start point X (meters)")] double x1,
        [Description("Start point Y (meters)")] double y1,
        [Description("End point X (meters)")] double x2,
        [Description("End point Y (meters)")] double y2,
        [Description("Arc direction: 1 = counter-clockwise, -1 = clockwise")] int direction = 1)
    {
        var info = await sta.InvokeLoggedAsync(nameof(AddArc), new { cx, cy, x1, y1, x2, y2, direction }, () => sketch.AddArc(cx, cy, x1, y1, x2, y2, direction));
        return JsonSerializer.Serialize(info);
    }

    [McpServerTool, Description("Draw an ellipse in the active sketch. All coordinates in meters.")]
    public async Task<string> AddEllipse(
        [Description("Center X (meters)")] double cx,
        [Description("Center Y (meters)")] double cy,
        [Description("Major-axis point X (meters)")] double majorX,
        [Description("Major-axis point Y (meters)")] double majorY,
        [Description("Minor-axis point X (meters)")] double minorX,
        [Description("Minor-axis point Y (meters)")] double minorY)
    {
        var info = await sta.InvokeLoggedAsync(nameof(AddEllipse), new { cx, cy, majorX, majorY, minorX, minorY }, () => sketch.AddEllipse(cx, cy, majorX, majorY, minorX, minorY));
        return JsonSerializer.Serialize(info);
    }

    [McpServerTool, Description("Draw a regular polygon in the active sketch. Coordinates in meters.")]
    public async Task<string> AddPolygon(
        [Description("Center X (meters)")] double cx,
        [Description("Center Y (meters)")] double cy,
        [Description("A point on the perimeter X (meters)")] double x,
        [Description("A point on the perimeter Y (meters)")] double y,
        [Description("Number of sides (minimum 3)")] int sides,
        [Description("True = inscribed (vertex on circle), False = circumscribed (edge tangent to circle)")] bool inscribed = true)
    {
        var info = await sta.InvokeLoggedAsync(nameof(AddPolygon), new { cx, cy, x, y, sides, inscribed }, () => sketch.AddPolygon(cx, cy, x, y, sides, inscribed));
        return JsonSerializer.Serialize(info);
    }

    [McpServerTool, Description("Insert sketch text at the given position in the active sketch.")]
    public async Task<string> AddText(
        [Description("Text anchor X (meters)")] double x,
        [Description("Text anchor Y (meters)")] double y,
        [Description("Text content to insert")] string text)
    {
        var info = await sta.InvokeLoggedAsync(nameof(AddText), new { x, y, text }, () => sketch.AddText(x, y, text));
        return JsonSerializer.Serialize(info);
    }
}
